src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,910 @@
+	}
+
+	if enc.numEncoded == 0 {
+		return ts.Datapoint{}, errNoEncodedDatapoints
+	}
+
+	return enc.lastEncodedDP, nil
+}
+
+// Len returns the length of the data stream.
+func (enc *Encoder) Len() int {
+	return enc.stream.Len()
+}
+
+func (enc *Encoder) encodeHeader() {
+	enc.encodeVarInt(currentEncodingSchemeVersion)
+	enc.encodeVarInt(uint64(enc.opts.ByteFieldDictionaryLRUSize()))
+	enc.encodeCustomSchemaTypes()
+}
+
+func (enc *Encoder) encodeCustomSchemaTypes() {
+	if len(enc.customFields) == 0 {
+		enc.encodeVarInt(0)
+		return
+	}
+
+	// Field numbers are 1-indexed so encoding the maximum field number
+	// at the beginning is equivalent to encoding the number of types
+	// we need to read after if we imagine that we're encoding a 1-indexed
+	// bitset where the position in the bitset encodes the field number (I.E
+	// the first value is the type for field number 1) and the values are
+	// the number of bits required to unique identify a custom type instead of
+	// just being a single bit (3 bits in the case of version 1 of the encoding
+	// scheme.)
+	maxFieldNum := enc.customFields[len(enc.customFields)-1].fieldNum
+	enc.encodeVarInt(uint64(maxFieldNum))
+
+	// Start at 1 because we're zero-indexed.
+	for i := 1; i <= maxFieldNum; i++ {
+		customTypeBits := uint64(cNotCustomEncoded)
+		for _, customField := range enc.customFields {
+			if customField.fieldNum == i {
+				customTypeBits = uint64(customField.fieldType)
+				break
+			}
+		}
+
+		enc.stream.WriteBits(
+			customTypeBits,
+			numBitsToEncodeCustomType)
+	}
+}
+
+func (enc *Encoder) encodeProto(m *dynamic.Message) error {
+	if err := enc.encodeCustomValues(m); err != nil {
+		return err
+	}
+	if err := enc.encodeProtoValues(m); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reset resets the encoder for reuse.
+func (enc *Encoder) Reset(
+	start time.Time,
+	capacity int,
+) {
+	enc.reset(start, capacity)
+}
+
+// SetSchema sets the encoders schema.
+// TODO(rartoul): Add support for changing the schema (and updating the ordering
+// of the custom encoded fields) on demand: https://github.com/m3db/m3/issues/1471
+func (enc *Encoder) SetSchema(schema *desc.MessageDescriptor) {
+	enc.resetSchema(schema)
+}
+
+func (enc *Encoder) reset(start time.Time, capacity int) {
+	enc.stream.Reset(enc.newBuffer(capacity))
+	enc.timestampEncoder = m3tsz.NewTimestampEncoder(
+		start, enc.opts.DefaultTimeUnit(), enc.opts)
+	enc.lastEncoded = nil
+	enc.lastEncodedDP = ts.Datapoint{}
+	enc.unmarshaled = nil
+
+	if enc.schema != nil {
+		enc.customFields = resetCustomFields(enc.customFields, enc.schema)
+	}
+
+	enc.hasEncodedFirstSetOfCustomValues = false
+	enc.closed = false
+	enc.numEncoded = 0
+}
+
+func (enc *Encoder) resetSchema(schema *desc.MessageDescriptor) {
+	enc.schema = schema
+	enc.customFields = resetCustomFields(enc.customFields, enc.schema)
+
+	enc.lastEncoded = dynamic.NewMessage(schema)
+	enc.unmarshaled = dynamic.NewMessage(schema)
```

Comment:
Is it possible to reuse dynamic.Message at all and just call reset here instead of allocating?

Reply:
If the schema doesn't change, I could just call `Reset`. To do that I would need an easy way to check if two schemas are equal, which I won't have until Haijun's code lands.

Added a comment and tracking here: https://github.com/m3db/m3/issues/1471